lib/Marlin/Marlin/src/feature/twibus.cpp

```diff
@@ -26,174 +26,216 @@
 
 #include "twibus.h"
 
-#include <Wire.h>
+FORCE_INLINE char hex_nybble(const uint8_t n) {
+  return (n & 0xF) + ((n & 0xF) < 10 ? '0' : 'A' - 10);
+}
+
+TWIBus twibus;
 
 TWIBus::TWIBus() {
-  #if I2C_SLAVE_ADDRESS == 0
-    Wire.begin();                  // No address joins the BUS as the master
-  #else
-    Wire.begin(I2C_SLAVE_ADDRESS); // Join the bus as a slave
-  #endif
   reset();
 }
 
 void TWIBus::reset() {
   buffer_s = 0;
   buffer[0] = 0x00;
+  read_buffer_available = 0;
+  read_buffer_pos = 0;
+}
+
+bool TWIBus::read_buffer_has_byte() {
+  return read_buffer_pos < read_buffer_available;
 }
 
-void TWIBus::address(const uint8_t adr) {
+uint8_t TWIBus::read_buffer_read_byte() {
+  if (!read_buffer_has_byte()) {
+    return 0;
+  }
+  return read_buffer[read_buffer_pos++];
+}
+
+bool TWIBus::address(const uint8_t adr) {
   if (!WITHIN(adr, 8, 127)) {
     SERIAL_ECHO_MSG("Bad I2C address (8-127)");
+    return false;
+  }
+
+  if (isRestrictedAddress(const uint8_t addr)) {
+    SERIAL_ECHO_MSG("Restricted I2C address.");
+    return false
   }
 
   addr = adr;
 
-  #if ENABLED(DEBUG_TWIBUS)
-    debug(PSTR("address"), adr);
-  #endif
+  debug(F("address"), adr);
+  return true;
+}
+
+bool TWIBus:isRestrictedAddress(uint8_t addr) {
+  switch (addr) {
+    case 0xa6, 0xae, 0xa7, 0xaf:
+      // EEPROM
+      return true;
+    case 0x22, 0x23:
+      // USBC
+      return true;
+     case 0x30, 0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37, 0x38, 0x39, 0x3A, 0x3B, 0x3C, 0x3D, 0x3E, 0x3F:
+      // IO Extender
+      return true;
+  }
+
+  return false;
 }
 
 void TWIBus::addbyte(const char c) {
   if (buffer_s >= COUNT(buffer)) return;
   buffer[buffer_s++] = c;
-  #if ENABLED(DEBUG_TWIBUS)
-    debug(PSTR("addbyte"), c);
-  #endif
+  debug(F("addbyte"), c);
 }
 
 void TWIBus::addbytes(char src[], uint8_t bytes) {
-  #if ENABLED(DEBUG_TWIBUS)
-    debug(PSTR("addbytes"), bytes);
-  #endif
+  debug(F("addbytes"), bytes);
   while (bytes--) addbyte(*src++);
 }
 
 void TWIBus::addstring(char str[]) {
-  #if ENABLED(DEBUG_TWIBUS)
-    debug(PSTR("addstring"), str);
-  #endif
+  debug(F("addstring"), str);
   while (char c = *str++) addbyte(c);
 }
 
 void TWIBus::send() {
-  #if ENABLED(DEBUG_TWIBUS)
-    debug(PSTR("send"), addr);
-  #endif
-
-  Wire.beginTransmission(I2C_ADDRESS(addr));
-  Wire.write(buffer, buffer_s);
-  Wire.endTransmission();
-
+  debug(F("send"), addr);
+  
+  i2c::Result ret = i2c::Transmit(&hi2c2, addr << 1, buffer, buffer_s, 100);
```

Comment:
First parameter of i2c::Transmit is a reference (not a pointer), so it should be without "&".